Updates CI to fix broken checks and builds caused by WASM introduction

ibeckermayer opened the pull request:
- Updates the `.github/workflows/lint-ui.yaml` and `.github/workflows/unit-tests-ui.yaml` GHA to use the Teleport buildbox container. This allows these CI runs to include a `Build WASM` step using [`wasm-pack`](https://github.com/rustwasm/wasm-pack), which is required for building the rust-based WASM module. This WASM module is required for the linter and UI tests to pass.
- Add `babel-plugin-transform-import-meta` and `babel-plugin-transform-vite-meta-env` to the babel config. This is necessary for `import.meta.url` and `import.meta.env` (respectively) to work in a nodejs environment (which is to say, for the `jest` tests to work).
- Changes all `from 'teleport/ironrdp/pkg'` to `from 'teleport/ironrdp/pkg/ironrdp'` It's not entirely clear to me why this is necessary, it's likely some discrepancy between how vite does things vs how nodejs wants things. This is an easy enough fix and it works (no need to get fancy with further babel magic).
- Updates README with some information on WASM.

Related PRs:
- https://github.com/gravitational/teleport.e/pull/2944
- https://github.com/gravitational/teleport/pull/28105
- https://github.com/gravitational/teleport/pull/35731

Commit:
Fix `yarn test`

Change all `from 'teleport/ironrdp/pkg'` to `from 'teleport/ironrdp/pkg/ironrdp'`
It's not entirely clear to me why this is necessary, it's some discrepancy between
how vite does things vs how nodejs wants things. This is easy enough and works.

Also add `babel-plugin-transform-import-meta` and `babel-plugin-transform-vite-meta-env`
to the babel config. This is necessary for the `import.meta.url` and `import.meta.env`
stuff to work.

Commit:
prepend `yarn build-wasm` to test and type-check

camscale commented on package.json:
```diff
@@ -45,4 +45,4 @@
     ]
   }
-}
+}
```
your editor seems to be removing newlines at the end of files. Can you fix that do it doesn't do that? Every line should end with a newline, including the last (see https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file)

camscale commented on web/packages/teleport/package.json:
```diff
@@ -28,6 +28,8 @@
+    "babel-plugin-transform-vite-meta-env": "^1.0.3"
   }
-}
+}
```
same here

ibeckermayer replied:
a80054fdfaaec1b3ad8eb44e632baca3b2173a2f

camscale commented on package.json:
```diff
@@ -13,13 +13,13 @@
     "storybook": "start-storybook -p 9002 -c web/.storybook -s web/.storybook/public",
     "storybook-smoke-test": "yarn storybook --ci --smoke-test",
-    "test": "jest",
+    "test": "yarn build-wasm && jest",
```
Adding a new build tool requirement is something that should be called out loudly. It seems rather hidden in this PR titled "Fix yarn test" where the PR description does not mention it either. I'm not sure of the relationship between the rust `wasm-pack` package/module and the yarn `build-wasm` command so I think that should be documented somewhere somehow too.

ibeckermayer replied:
These were removed in favor of running them as a separate step in the CI pipeline https://github.com/gravitational/teleport/pull/35401/commits/9a3579c00fd89c4253d93d545ea06d1bc75820cd

This is because we don't necessarily want to ensure wasm is built before _every_ run of `yarn test`, e.g. sometimes we just want to do `yarn test <specific test which may not even rely on wasm modules>

camscale commented on web/packages/teleport/package.json:
```diff
@@ -28,6 +28,8 @@
     "jest-canvas-mock": "^2.3.1",
-    "ts-loader": "^9.4.2"
+    "ts-loader": "^9.4.2",
+    "babel-plugin-transform-import-meta": "^2.2.0",
```
Not sure if these deps were added manually or automatically, but if manually I find it is best to keep lists like this in alphabetical order. It makes it easy to know when you have duplicates (or more accurately you don't get duplicates because it is obvious) and it helps manage merge conflicts when things get added at different places rather than all at the end. If these were added automatically by some command, please ignore this comment.

ibeckermayer replied:
a80054fdfaaec1b3ad8eb44e632baca3b2173a2f

Commit:
Adds back newlines at EOF, alphabetizes devDependencies

Commit:
remove yarn build-wasm &&, add a yarn build-wasm step to UI testing/linting workflows, run UI testing/linting workflows on buildbox

Commit:
Tell prettier to ignore generated WASM glue

Commit:
Tell eslint to ignore generated WASM glue

Commit:
Adds the --init option to the unit-tests-ui GHA and adds an explanation in the relevant test. (A link to the explanation will be added near the --init option in the next commit.)

Commit:
Add link to explanation for --init

Commit:
Update the README with some information on WASM

Commit:
Release v15.0.0-dev.isaiah.1

Commit:
Merge branch 'ironrdp' into ironrdp-fix-jest-tests

Commit:
Release v15.0.0-dev.isaiah.2

camscale commented on Makefile:
```diff
@@ -11,7 +11,7 @@
 #   Pre-releases:      "1.0.0-alpha.1", "1.0.0-beta.2", "1.0.0-rc.3"
 #   Master/dev branch: "1.0.0-dev"
-VERSION=15.0.0-dev
+VERSION=15.0.0-dev.isaiah.1
```
Ensure this change does not get merged please.

What I usually do is create a `camh/test/foo` branch for `camh/foo` changes and run the test build from there - it keeps the PR branch clean and you just switch back to it and delete the test branch to clean up. The tags still hang around though. We should consider cleaning up dev tags every now and then.

ibeckermayer replied:
Good call, I'll do that in the future. Should be polished for review now.

camscale reviewed (commented):
It's a bit hard to review this with all the changes from `make update-version`. I've suggested a way to avoid that, but please ensure you don't merge this with the results of running that.

Commit:
adds ironrdp/pkg to files to ignore in the calculate_sha find command

Commit:
Release v15.0.0-dev.isaiah.3

Commit:
Add some stdout finesse to build.assets/build-webassets-if-changed.sh

Commit:
Use latest e hash with CI update
This e hash is on a branch that hasn't been merged yet,
if you find this commit in master there's been a mistake.

Commit:
Release v15.0.0-dev.isaiah.4

Commit:
Revert "Use latest e hash with CI update"

This reverts commit 8f82c4d548f59990fa129ffa91542113147b5719.

Commit:
Updating e-ref to "Installs wasm-pack in build-mac.yaml"
This is for testing purposes, this e-ref has not been merged into
teleport.e's master branch yet.

Commit:
Release v15.0.0-dev.isaiah.5

For testing build-darwin-amd64

Commit:
Updating e to Add build-wasm to enterprise package.json

Commit:
Release v15.0.0-dev.isaiah.6

For testing build-darwin-amd64

Commit:
Updating e-ref to Installs wasm-pack in build-mac-universal.yaml

Commit:
Release v15.0.0-dev.isaiah.7

For testing build-darwin-amd64

Commit:
Make release-arm rely on webassets similar to centos7, so that we can build the webassets with buildbox-node

Commit:
Release v15.0.0-dev.isaiah.8

For testing build-linux-arm

Commit:
Add scl enable ${DEVTOOLSET} before wasm-pack install
Otherwise we try to use an ancient compiler which can't compile
one of wasm-pack's dependencies.

Commit:
Release v15.0.0-dev.isaiah.9

For testing build-linux-arm64

Commit:
split rustup-install-target-toolchain into itself and rustup-set-version, change e-ref to point at e1fe596ea93e0fed0a2551c26c81ed7d8e035e78 which uses rustup-set-version

Commit:
Release v15.0.0-dev.isaiah.10

For testing build-darwin-amd64

Commit:
release-arm also using release-unix-preserving-webassets

Commit:
Release v15.0.0-dev.isaiah.11

For testing build-linux-arm

Commit:
setting e-ref to aa84b4af7a624d1d9dab2242cbed6faa5bd8117a

Commit:
Release v15.0.0-dev.isaiah.12

For testing build-darwin-amd64

Commit:
Have release-arm mimic release-arm64 for now. this is ugly but in theory should get the job done

Commit:
Release v15.0.0-dev.isaiah.13

For testing build-linux-arm

Commit:
Attempts to parameterize release target so that we can optionally pass release-unix-preserving-webassets in the case of release-arm

Commit:
Release v15.0.0-dev.isaiah.14

For testing build-linux-arm

Commit:
Release v15.0.0-dev.isaiah.15

Checking all builds succeed

Commit:
Release v15.0.0-dev.isaiah.17

Commit:
fix linter errs

Commit:
reverting changes from build test runs

Commit:
removing superfluous ** in ignore files

ibeckermayer commented on .eslintignore:
```diff
@@ -1,2 +1,5 @@
 .eslintrc.js
+
+# Ignore WASM generated files:
+web/packages/teleport/src/ironrdp/pkg/**
```
It needn't be in the gitignore because wasm-pack always generates this directory with a gitignore inside with

```
*
```

I did remove the ** from here and .prettierignore and everything still works.

ibeckermayer commented:
In theory every build pipeline is passing now. In practice I had one fail on me due to a GHA error:

<img width="991" alt="image" src="https://github.com/gravitational/teleport/assets/13578537/ca7373ee-d3be-4ba5-b94a-9adef598f7be">

https://drone.platform.teleport.sh/gravitational/teleport/31857/10/1
https://github.com/gravitational/teleport.e/actions/runs/7225580694

However I have another run where that succeeded and I haven't made relevant changes since:
<img width="1000" alt="image" src="https://github.com/gravitational/teleport/assets/13578537/5c29778f-49b3-4b1e-a749-5ac27afa0508">
https://drone.platform.teleport.sh/gravitational/teleport/31839/6/2
https://github.com/gravitational/teleport.e/actions/runs/7216993875

Other runs are failing for reasons unrelated to what I'm doing.

Therefore once the related `e` changes have merged (https://github.com/gravitational/teleport.e/pull/2944) and this PR's e-ref updated to the merged commit, this should be good to go.

zmb3 commented on Makefile:
```diff
@@ -11,7 +11,7 @@
 #   Pre-releases:      "1.0.0-alpha.1", "1.0.0-beta.2", "1.0.0-rc.3"
 #   Master/dev branch: "1.0.0-dev"
-VERSION=15.0.0-dev
+VERSION=15.0.0-dev.isaiah.17
```
Don't forget to revert this.

ibeckermayer replied:
e63a6e2c59025fae385617c86d82742e6d8fc525

zmb3 commented on e:
Don't forget to update this after the `e` PR merges.

zmb3 commented on web/packages/teleterm/src/agentCleanupDaemon/agentCleanupDaemon.test.ts:
```diff
@@ -80,6 +80,13 @@ describe('agentCleanupDaemon', () => {
+      // In typical UNIX environments, zombies exist only momentarily and are cleaned up by the init process (https://en.wikipedia.org/wiki/Init),
+      // however in Docker there is no init process by default (https://blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem/),
+      // so zombie processes end up sticking around until the container is stopped. Hence in order for this test to pass in a `docker run`,
+      // we need to add the `--init` flag (https://docs.docker.com/engine/reference/run/#specify-an-init-process).
```
Where is said `--init` flag?

ibeckermayer replied:
It's used here:

https://github.com/gravitational/teleport/pull/35401/files#diff-6c099f0fc9d8f8a6677998cdb8f0f619e662a887bf91adbe15d7a2dc5af57ebbR22

zmb3 replied:
Gotcha. Any reason why the large comment explaining why the flag is necessary is so far away from where the flag is used? Would it make more sense to move this reasoning over there?

ibeckermayer replied:
I understood the need for that flag in close relation to the behavior of this specific test, so wrote it here in order to codify that. IMO I would need to link to this test in order to properly document why `init` was added over in `unit-tests-ui.yaml`, so I figured I may as well leave the explanation where it is and [just link to explanation + test](https://github.com/gravitational/teleport/pull/35401/files#diff-6c099f0fc9d8f8a6677998cdb8f0f619e662a887bf91adbe15d7a2dc5af57ebbR20-R21).

camscale commented on e:
the commit you are updating `e` to point to is not on master on teleport.e - this should not be merged to teleport master

camscale commented on web/README.md:
```diff
@@ -117,6 +117,14 @@ requests to the given target.
 
+### WASM
+
+The web UI includes a WASM modul built from a Rust codebase located in `packages/teleport/src/ironrdp`.
```
s/modul/module/

ibeckermayer replied:
72f0aac4b8637a1263c53e72d203fc598f25bb74

camscale commented on build.assets/Makefile:
```diff
@@ -82,6 +82,8 @@ REQUIRE_HOST_ARCH = $(if $(filter-out $(ARCH),$(RUNTIME_ARCH)),$(error Cannot cr
 # needed on CI as they start with a fresh state each time.
 PULL_ON_CI = $(if $(filter $(CI),true),docker inspect --type=image $(1) >/dev/null 2>&1 || docker pull $(1) || true)
 
+RELEASE_TARGET ?= release
```
The comments about release-unix-preserving-webassets etc should probably be attached to this line. It is far from where it is used so should have some comment at least.

ibeckermayer replied:
See 1de16ff8d8b47eb3abcbe89c57729ce328d56fa1

camscale commented on web/README.md:
```diff
@@ -117,6 +117,14 @@ requests to the given target.
 
+### WASM
+
+The web UI includes a WASM modul built from a Rust codebase located in `packages/teleport/src/ironrdp`.
```
This PR does not include anything in `web/packages/teleport/src/ironrdp` and there is nothing there on master. I'm a little confused how this PR works without this source code. Does this README change belong in a different PR?

ibeckermayer replied:
This PR is being merged into the `ironrdp` branch, which is the big feature branch for the IronRDP client refactor:

https://github.com/gravitational/teleport/pull/26874

Commit:
removing remaining .isaiah.17's

ibeckermayer commented on Makefile:
```diff
@@ -11,7 +11,7 @@
 #   Pre-releases:      "1.0.0-alpha.1", "1.0.0-beta.2", "1.0.0-rc.3"
 #   Master/dev branch: "1.0.0-dev"
-VERSION=15.0.0-dev
+VERSION=15.0.0-dev.isaiah.17
```
e63a6e2c59025fae385617c86d82742e6d8fc525

Commit:
bash wizardry

ibeckermayer commented on build.assets/build-webassets-if-changed.sh:
```diff
@@ -10,15 +10,21 @@ ROOT_PATH="$(cd "$(dirname "$0")/.." && pwd -P)"
+function print_for_user() {
+  local script_name
+  script_name=$(basename "$0")
+  echo "$script_name: $1"
```
🧙‍♂️

3103c30f844a4b0faab5dfd7a54e250e01347b64

Commit:
Adds comment to RELEASE_TARGET

Commit:
typo

camscale reviewed (approved):
LGTM. I'm not sure how this gets into master in the end and how long that takes. It's quite possible we'll make changes to the same makefile targets as we clean up, so these changes will conflict when the ironrdp branch is being merged to master.

ibeckermayer commented:
> LGTM. I'm not sure how this gets into master in the end and how long that takes. It's quite possible we'll make changes to the same makefile targets as we clean up, so these changes will conflict when the ironrdp branch is being merged to master.

May god have mercy on our souls

Commit:
Update e ref

Commit:
Merge branch 'ironrdp' into ironrdp-fix-jest-tests
